llvm/lib/CodeGen/WasmEHPrepare.cpp

```diff
@@ -113,19 +116,41 @@ class WasmEHPrepare : public FunctionPass {
   bool prepareEHPads(Function &F);
   void prepareEHPad(BasicBlock *BB, bool NeedPersonality, unsigned Index = 0);
 
+public:
+  WasmEHPrepareImpl() = default;
+  WasmEHPrepareImpl(Type *LPadContextTy_) : LPadContextTy(LPadContextTy_) {}
+  bool runOnFunction(Function &F);
+};
+
+class WasmEHPrepare : public FunctionPass {
+  WasmEHPrepareImpl P;
+
 public:
   static char ID; // Pass identification, replacement for typeid
 
   WasmEHPrepare() : FunctionPass(ID) {}
   bool doInitialization(Module &M) override;
-  bool runOnFunction(Function &F) override;
+  bool runOnFunction(Function &F) override { return P.runOnFunction(F); }
 
   StringRef getPassName() const override {
     return "WebAssembly Exception handling preparation";
   }
 };
+
 } // end anonymous namespace
 
+PreservedAnalyses WasmEHPreparePass::run(Function &F,
+                                         FunctionAnalysisManager &) {
+  auto &Context = F.getContext();
```

Comment:
I'm not super familiar with the new pass manager, but this new code seems to be the same as what's in `WasmEHPrepare::doInitialization()` below. Are they both necessary?

Reply:
Then we should delete the current `doInitialization`. But then where do we do stuff that was only possible in `doInitialization` before, like things listed here? https://llvm.org/docs/WritingAnLLVMPass.html#the-doinitialization-module-method